fetch_raw_data.py

```diff
@@ -0,0 +1,247 @@
+                    flow_id, raw_runs, raw_export_log_file=raw_runs_log_file, ignore_archives=True)
+            except FileNotFoundError:
+                log.info(f"File '{raw_runs_path}' not found, will fetch all runs from the Rapid Pro server for flow '{flow}'")
+                raw_runs = rapid_pro.get_raw_runs_for_flow_id(flow_id, raw_export_log_file=raw_runs_log_file)
+
+        # Fetch the latest contacts from Rapid Pro.
+        with open(contacts_log_path, "a") as raw_contacts_log_file:
+            raw_contacts = rapid_pro.update_raw_contacts_with_latest_modified(raw_contacts,
+                                                                              raw_export_log_file=raw_contacts_log_file)
+
+        # Convert the runs to TracedData.
+        traced_runs = rapid_pro.convert_runs_to_traced_data(
+            user, raw_runs, raw_contacts, phone_number_uuid_table, rapid_pro_source.test_contact_uuids)
+
+        if flow in rapid_pro_source.activation_flow_names:
+            label_somalia_operator(user, traced_runs, phone_number_uuid_table)
+
+        log.info(f"Saving {len(raw_runs)} raw runs to {raw_runs_path}...")
+        with open(raw_runs_path, "w") as raw_runs_file:
+            json.dump([run.serialize() for run in raw_runs], raw_runs_file)
+        log.info(f"Saved {len(raw_runs)} raw runs")
+
+        log.info(f"Saving {len(traced_runs)} traced runs to {traced_runs_output_path}...")
+        IOUtils.ensure_dirs_exist_for_file(traced_runs_output_path)
+        with open(traced_runs_output_path, "w") as traced_runs_output_file:
+            TracedDataJsonIO.export_traced_data_iterable_to_jsonl(traced_runs, traced_runs_output_file)
+        log.info(f"Saved {len(traced_runs)} traced runs")
+
+    log.info(f"Saving {len(raw_contacts)} raw contacts to file '{raw_contacts_path}'...")
+    with open(raw_contacts_path, "w") as raw_contacts_file:
+        json.dump([contact.serialize() for contact in raw_contacts], raw_contacts_file)
+    log.info(f"Saved {len(raw_contacts)} contacts")
+    
+    
+def fetch_from_gcloud_bucket(google_cloud_credentials_file_path, raw_data_dir, gcloud_source):
+    log.info("Fetching data from a gcloud bucket...")
+    for blob_url in gcloud_source.activation_flow_urls + gcloud_source.survey_flow_urls:
+        flow = blob_url.split("/")[-1]
+
+        traced_runs_output_path = f"{raw_data_dir}/{flow}"
+        if os.path.exists(traced_runs_output_path):
+            log.info(f"File '{traced_runs_output_path}' for flow '{flow}' already exists; skipping download")
+            continue
+        
+        log.info(f"Saving '{flow}' to file '{traced_runs_output_path}'...")
+        with open(traced_runs_output_path, "wb") as traced_runs_output_file:
+            google_cloud_utils.download_blob_to_file(
+                google_cloud_credentials_file_path, blob_url, traced_runs_output_file)
+
+
+def fetch_from_shaqadoon_csv(user, google_cloud_credentials_file_path, raw_data_dir, phone_number_uuid_table,
```

Comment:
ditto although it doesn't hurt keeping it too.

Reply:
This was named back when we thought Shaqodoon would send us data in the same format every time. It turns out they don't, and so this has really become a standardised format that's easy for us to translate Shaqadoon's data into and that's easy for the pipeline to ingest. If we had to recover data from a different, failing MNO, this *might* be useful for that too. So I would suggest keeping it for this pipeline, in part because it might be useful and in part so we don't have to handle the complexity of deciding which parts of pipelines to copy in future. However, I suggest renaming it to "fetch_from_recovery_csv" to reflect its more recent, broader use.

Reply:
I'll rename in the next PR to make reviewing easier

Reply:
https://github.com/AfricasVoices/Project-COVID19/pull/2